GitCommands/Git/Executable.cs

```diff
@@ -212,14 +212,7 @@ public StreamReader StandardError
                         throw new InvalidOperationException("Process was not created with redirected output.");
                     }
 
-                    if (!_throwOnErrorExit)
-                    {
-                        return _process.StandardError;
-                    }
-
-                    _emptyStream ??= new();
-                    _emptyReader ??= new(_emptyStream);
-                    return _emptyReader;
+                    return _process.StandardError;
```

Comment:
No, the StandardError has already been consumed and is stored in the ExternalOperationException.
If this really works, i.e. StandardError could be read again, `_emptyStream` and `_emptyReader` should be removed. (I should have used a single `Lazy<>`.)

Reply:
If the exception is raised, we should never get here.
If there is a way, then the check should include exit status

Reply:
```
Entering 'Externals/EasyHook'
Entering 'Externals/Git.hub'
Removing Git.hub/bin/
Removing Git.hub/obj/
Entering 'Externals/ICSharpCode.TextEditor'
Removing Project/bin/
Removing Project/obj/
Entering 'Externals/conemu-inside'
Removing ConEmuWinForms/bin/
Removing ConEmuWinForms/obj/
Removing Externals/NetSpell.SpellChecker/bin/
Removing Externals/NetSpell.SpellChecker/obj/
Removing artifacts/
MSBuild version 17.3.2+561848881 for .NET
  Determining projects to restore...
  Restored C:\git\gitextensions\GitExtUtils\GitExtUtils.csproj (in 769 ms).
  Restored C:\git\gitextensions\GitCommands\GitCommands.csproj (in 284 ms).
  Restored C:\git\gitextensions\Plugins\AutoCompileSubmodules\GitExtensions.Plugins.AutoCompileSubmodules.csproj (in 1.12 sec).
  Restored C:\git\gitextensions\Plugins\BackgroundFetch\GitExtensions.Plugins.BackgroundFetch.csproj (in 1.12 sec).
  Restored C:\git\gitextensions\Plugins\BuildServerIntegration\AppVeyorIntegration\AppVeyorIntegration.csproj (in 1.12 sec).
  Restored C:\git\gitextensions\GitExtensions\GitExtensions.csproj (in 1.12 sec).
  Restored C:\git\gitextensions\Plugins\Bitbucket\GitExtensions.Plugins.Bitbucket.csproj (in 1.12 sec).
  Restored C:\git\gitextensions\Plugins\BuildServerIntegration\JenkinsIntegration\JenkinsIntegration.csproj (in 26 ms).
  Restored C:\git\gitextensions\Plugins\CreateLocalBranches\GitExtensions.Plugins.CreateLocalBranches.csproj (in 31 ms).
  Restored C:\git\gitextensions\Plugins\BuildServerIntegration\AzureDevOpsIntegration\AzureDevOpsIntegration.csproj (in 49 ms).
  Restored C:\git\gitextensions\Plugins\BuildServerIntegration\TeamCityIntegration\TeamCityIntegration.csproj (in 34 ms).
  Restored C:\git\gitextensions\Plugins\DeleteUnusedBranches\GitExtensions.Plugins.DeleteUnusedBranches.csproj (in 44 ms).
  Restored C:\git\gitextensions\Plugins\Gource\GitExtensions.Plugins.Gource.csproj (in 101 ms).
  Restored C:\git\gitextensions\Plugins\GitUIPluginInterfaces\GitUIPluginInterfaces.csproj (in 64 ms).
  Restored C:\git\gitextensions\Plugins\GitHub3\GitExtensions.Plugins.GitHub3.csproj (in 29 ms).
  Restored C:\git\gitextensions\Plugins\GitFlow\GitExtensions.Plugins.GitFlow.csproj (in 25 ms).
  Restored C:\git\gitextensions\Plugins\FindLargeFiles\GitExtensions.Plugins.FindLargeFiles.csproj (in 34 ms).
  Restored C:\git\gitextensions\Plugins\Statistics\GitStatistics\GitExtensions.Plugins.GitStatistics.csproj (in 32 ms).
  Restored C:\git\gitextensions\Plugins\Statistics\GitImpact\GitExtensions.Plugins.GitImpact.csproj (in 27 ms).
  Restored C:\git\gitextensions\IntegrationTests\UI.IntegrationTests\UI.IntegrationTests.csproj (in 1.56 sec).
  Restored C:\git\gitextensions\UnitTests\GitExtUtils.Tests\GitExtUtils.Tests.csproj (in 1.56 sec).
  Restored C:\git\gitextensions\UnitTests\Plugins\BuildServerIntegration\AzureDevOpsIntegration.Tests\AzureDevOpsIntegration.Tests.csproj (in 327 ms).
  Restored C:\git\gitextensions\UnitTests\GitUI.Tests\GitUI.Tests.csproj (in 316 ms).
  Restored C:\git\gitextensions\UnitTests\Plugins\BuildServerIntegration\AppVeyorIntegration.Tests\AppVeyorIntegration.Tests.csproj (in 315 ms).
  Restored C:\git\gitextensions\Plugins\ProxySwitcher\GitExtensions.Plugins.ProxySwitcher.csproj (in 20 ms).
  Restored C:\git\gitextensions\UnitTests\Plugins\DeleteUnusedBranches.Tests\DeleteUnusedBranches.Tests.csproj (in 389 ms).
  Restored C:\git\gitextensions\Plugins\ReleaseNotesGenerator\GitExtensions.Plugins.ReleaseNotesGenerator.csproj (in 38 ms).
  Restored C:\git\gitextensions\UnitTests\ResourceManager.Tests\ResourceManager.Tests.csproj (in 72 ms).
  Restored C:\git\gitextensions\UnitTests\CommonTestUtils\CommonTestUtils.csproj (in 89 ms).
  Restored C:\git\gitextensions\UnitTests\GitCommands.Tests\GitCommands.Tests.csproj (in 138 ms).
  Restored C:\git\gitextensions\UnitTests\Plugins\ReleaseNotesGenerator.Tests\ReleaseNotesGenerator.Tests.csproj (in 163 ms).
  Restored C:\git\gitextensions\ResourceManager\ResourceManager.csproj (in 53 ms).
  Restored C:\git\gitextensions\TranslationApp\TranslationApp.csproj (in 145 ms).
  Restored C:\git\gitextensions\UnitTests\Plugins\GitUIPluginInterfaces.Tests\GitUIPluginInterfaces.Tests.csproj (in 201 ms).
  Restored C:\git\gitextensions\Externals\ICSharpCode.TextEditor\Project\ICSharpCode.TextEditor.csproj (in 2 ms).
  Restored C:\git\gitextensions\Externals\NetSpell.SpellChecker\SpellChecker.csproj (in 11 ms).
  Restored C:\git\gitextensions\UnitTests\BugReporter.Tests\BugReporter.Tests.csproj (in 163 ms).
  Restored C:\git\gitextensions\Plugins\JiraCommitHintPlugin\GitExtensions.Plugins.JiraCommitHintPlugin.csproj (in 219 ms).
  Restored C:\git\gitextensions\Externals\Git.hub\Git.hub\Git.hub.csproj (in 24 ms).
  Restored C:\git\gitextensions\Externals\conemu-inside\ConEmuWinForms\ConEmuWinForms.csproj (in 22 ms).
  Restored C:\git\gitextensions\GitUI\GitUI.csproj (in 53 ms).
  Restored C:\git\gitextensions\BugReporter\BugReporter.csproj (in 17 ms).
  Restored C:\git\gitextensions\IntegrationTests\BugReporter.IntegrationTests\BugReporter.IntegrationTests.csproj (in 88 ms).
  SpellChecker -> C:\git\gitextensions\Externals\NetSpell.SpellChecker\bin\Debug\net6.0-windows\SpellChecker.dll
  Git.hub -> C:\git\gitextensions\Externals\Git.hub\Git.hub\bin\Debug\netstandard2.0\Git.hub.dll
C:\git\gitextensions\Externals\conemu-inside\ConEmuWinForms\ConEmuControl.cs(59,77): warning VSTHRD110: Observe the awaitable result of this method call by awaiting it, assigning to a variable, or passing it to another method. [C:\git\gitextensions\Externals\conemu-inside\ConEmuWinForms\ConEmuWinForms.csproj]
  ConEmuWinForms -> C:\git\gitextensions\Externals\conemu-inside\ConEmuWinForms\bin\net6.0-windows\ConEmuWinForms.dll
  GitExtUtils -> C:\git\gitextensions\artifacts\Debug\bin\GitExtUtils\net6.0-windows\GitExtUtils.dll
  GitUIPluginInterfaces -> C:\git\gitextensions\artifacts\Debug\bin\GitUIPluginInterfaces\net6.0-windows\GitUIPluginInterfaces.dll
C:\git\gitextensions\GitCommands\Git\Executable.cs(67,35): error CS0649: Field 'Executable.ProcessWrapper._emptyStream' is never assigned to, and will always have its default value null [C:\git\gitextensions\GitCommands\GitCommands.csproj]
C:\git\gitextensions\GitCommands\Git\Executable.cs(68,35): error CS0649: Field 'Executable.ProcessWrapper._emptyReader' is never assigned to, and will always have its default value null [C:\git\gitextensions\GitCommands\GitCommands.csproj]

Build FAILED.

C:\git\gitextensions\Externals\conemu-inside\ConEmuWinForms\ConEmuControl.cs(59,77): warning VSTHRD110: Observe the awaitable result of this method call by awaiting it, assigning to a variable, or passing it to another method. [C:\git\gitextensions\Externals\conemu-inside\ConEmuWinForms\ConEmuWinForms.csproj]
C:\git\gitextensions\GitCommands\Git\Executable.cs(67,35): error CS0649: Field 'Executable.ProcessWrapper._emptyStream' is never assigned to, and will always have its default value null [C:\git\gitextensions\GitCommands\GitCommands.csproj]
C:\git\gitextensions\GitCommands\Git\Executable.cs(68,35): error CS0649: Field 'Executable.ProcessWrapper._emptyReader' is never assigned to, and will always have its default value null [C:\git\gitextensions\GitCommands\GitCommands.csproj]
    1 Warning(s)
    2 Error(s)

Time Elapsed 00:00:18.29
C:\git\gitextensions\scripts\..\artifacts
Untestable Commit.  Cannot find build output


```

Reply:
[Bisect-Test.txt](https://github.com/gitextensions/gitextensions/files/9791957/Bisect-Test.txt)

Rename to .ps1 file

Run in powershell
```ps
.\scripts\Bisect-Test.ps1  -Build -Run >Log.txt
```

Reply:
![image](https://user-images.githubusercontent.com/3629489/195986230-d266d358-42aa-459c-9390-1948ca5aab4c.png)
```cs
private MemoryStream? _emptyStream = null;
private StreamReader? _emptyReader = null;
```

Reply:
Digging into further not used but in dispose so...
```patch
-64,8 +64,6 @@ private sealed class ProcessWrapper : IProcess
             private readonly bool _redirectOutput;
             private readonly bool _throwOnErrorExit;
 
-            private MemoryStream? _emptyStream;
-            private StreamReader? _emptyReader;
 
             private bool _disposed;
 

@@ -254,9 +252,6 @@ public void Dispose()
                 _process.Dispose();
 
                 _logOperation.NotifyDisposed();
-
-                _emptyReader?.Dispose();
-                _emptyStream?.Dispose();
             }
         }
 ```

Reply:
I should have set this to draft, was in a hurry and there was a build error. Rather submitted the PR than just have a discussion, the error was clear to me - not the solution though.
Therefore hiding the posts about the build error, forced push the cleanup.
Also changed the redirect check.

I cannot see how StandardError can be accessed if the exception is thrown. The ExitCode could be read similar to how it is done in OnProcessExit(), but then the disposed handling (see comment in OnProcessExit()) maybe need to be handled, so I kept it simple.

StandardError is only used in a few situations, mostly from AllOutput. That should be changed to be StandardOutput in most situations, in GetCommitVerificationMessage (as well as in SystemEncodingReader.cs as a sample check) it should be StandardError though.

Reply:
The exception isn't being thrown.  The error code is 0.  Command ran fine.  Just because there is output in error stream does not mean it is an error.  
![image](https://user-images.githubusercontent.com/3629489/196005893-b700b859-e1cf-43e9-bf99-a4fe2b34fc85.png)

Reply:
This is the core error: stderr was not populated unless was set to false
This command seem to be the only situation where this information is used.

Will push a test update, then revert to only handle this error and cleanup stderr usage in another PR